File: apps/desktop/lib/screens/audit_log_screen.dart
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../design/nano_tokens.dart';
import '../models/models.dart';
import '../providers/app_provider.dart';
import '../utils/format.dart';
import '../widgets/nano/nano_card.dart';
import '../widgets/nano/nano_primitives.dart';

/// Full audit log of the active server's recent command activity.
///
/// Wires the real `GET /api/audit/recent` feed through
/// [AppProvider.fetchRecentActivity]; rows mirror the "最近活动" list in the
/// Android drawer/dashboard mockup (typed icon, mono command type, relative
/// time, params, user) but render the complete list with pull-to-refresh.
class AuditLogScreen extends StatefulWidget {
  const AuditLogScreen({super.key});

  @override
  State<AuditLogScreen> createState() => _AuditLogScreenState();
}

class _AuditLogScreenState extends State<AuditLogScreen> {
  bool _loading = true;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) => _load());
  }

  Future<void> _load() async {
    final provider = context.read<AppProvider>();
    await provider.fetchRecentActivity(limit: 200);
    if (mounted) setState(() => _loading = false);
  }

  @override
  Widget build(BuildContext context) {
    final t = context.nano;
    final provider = context.watch<AppProvider>();
    final entries = provider.recentActivity();

    return Scaffold(
      backgroundColor: t.bg,
      body: SafeArea(
        bottom: false,
        child: Column(
          children: [
            _AuditHeader(),
            Expanded(
              child: RefreshIndicator(
                onRefresh: _load,
                color: t.accent,
                backgroundColor: t.card,
                child: _loading && entries.isEmpty
                    ? ListView(
                        children: [
                          const SizedBox(height: 120),
                          Center(
                            child: CircularProgressIndicator(
                                color: t.accent, strokeWidth: 2.5),
                          ),
                        ],
                      )
                    : entries.isEmpty
                        ? _Empty()
                        : ListView(
                            padding: const EdgeInsets.fromLTRB(16, 4, 16, 96),
                            children: [
                              NanoSectionLabel(
                                'audit.entriesCount'.tr(
                                    namedArgs: {'n': '${entries.length}'}),
                              ),
                              NanoCard(
                                child: Column(
                                  children: [
                                    for (var i = 0; i < entries.length; i++)
                                      _AuditRow(
                                        entry: entries[i],
                                        divider: i < entries.length - 1,
                                      ),
                                  ],
                                ),
                              ),
                            ],
                          ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _AuditHeader extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final t = context.nano;
    return Padding(
      padding: EdgeInsets.fromLTRB(4, t.isIOS ? 12 : 8, 16, 8),
      child: Row(
        children: [
          IconButton(
            icon: Icon(
                t.isIOS
                    ? Icons.arrow_back_ios_new_rounded
                    : Icons.arrow_back_rounded,
                color: t.isIOS ? t.accent : t.fg,
                size: t.isIOS ? 20 : 24),
            onPressed: () => Navigator.pop(context),
          ),
          Expanded(
            child: Text('audit.title'.tr(),
                style: TextStyle(
                    fontSize: t.isIOS ? 22 : 22,
                    fontWeight: t.isIOS ? FontWeight.w700 : FontWeight.w500,
                    letterSpacing: t.displayTracking,
                    color: t.fg)),
          ),
        ],
      ),
    );
  }
}

class _Empty extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final t = context.nano;
    return ListView(
      padding: const EdgeInsets.fromLTRB(16, 80, 16, 16),
      children: [
        Icon(Icons.history_rounded, color: t.fg4, size: 34),
        const SizedBox(height: 12),
        Center(
          child: Text('audit.empty'.tr(),
              style: TextStyle(
                  color: t.fg2, fontSize: 15, fontWeight: FontWeight.w500)),
        ),
        const SizedBox(height: 4),
        Center(
          child: Text('audit.emptySub'.tr(),
              style: TextStyle(color: t.fg4, fontSize: 12.5)),
        ),
      ],
    );
  }
}

class _AuditRow extends StatelessWidget {
  final AuditEntry entry;
  final bool divider;
  const _AuditRow({required this.entry, required this.divider});

  IconData get _icon {
    switch (entry.type) {
      case 'shell.exec':
        return Icons.terminal_rounded;
      case 'service.restart':
      case 'docker.restart':
        return Icons.refresh_rounded;
      case 'system.reboot':
        return Icons.power_settings_new_rounded;
      default:
        return Icons.history_rounded;
    }
  }

  @override
  Widget build(BuildContext context) {
    final t = context.nano;
    final tone = entry.ok ? t.ok : t.crit;
    // Prefer the typed params summary; fall back to target / hostname.
    final pm = entry.paramsMap;
    final detail = pm.isNotEmpty
        ? pm.entries.map((e) => '${e.key}=${e.value}').join(' ')
        : (entry.target.isNotEmpty ? entry.target : entry.agentHostname);

    return NanoListRow(
      divider: divider,
      crossAxis: CrossAxisAlignment.start,
      leading: Container(
        width: 36,
        height: 36,
        decoration: BoxDecoration(
          color: tone.withValues(alpha: 0.16),
          borderRadius: BorderRadius.circular(18),
        ),
        child: Icon(_icon, color: tone, size: 16),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            crossAxisAlignment: CrossAxisAlignment.baseline,
            textBaseline: TextBaseline.alphabetic,
            children: [
              Flexible(
                child: NanoMono(entry.type,
                    size: 13,
                    weight: FontWeight.w500,
                    overflow: TextOverflow.ellipsis),
              ),
              const SizedBox(width: 6),
              Text(Fmt.ago(entry.at),
                  style: TextStyle(fontSize: 11, color: t.fg4)),
            ],
          ),
          if (detail.isNotEmpty) ...[
            const SizedBox(height: 2),
            NanoMono(detail,
                size: 11.5, color: t.fg3, overflow: TextOverflow.ellipsis),
          ],
          if (entry.error.isNotEmpty) ...[
            const SizedBox(height: 2),
            Text(entry.error,
                style: TextStyle(fontSize: 11.5, color: t.crit, height: 1.3),
                maxLines: 2,
                overflow: TextOverflow.ellipsis),
          ],
          const SizedBox(height: 4),
          Row(
            children: [
              if (entry.user.isNotEmpty)
                NanoBadge(entry.user, icon: Icons.person_outline_rounded),
              if (entry.user.isNotEmpty && entry.agentHostname.isNotEmpty)
                const SizedBox(width: 6),
              if (entry.agentHostname.isNotEmpty)
                NanoBadge(entry.agentHostname, mono: true),
              const Spacer(),
              if (entry.durationMs > 0)
                Text('${entry.durationMs}ms',
                    style: TextStyle(
                        fontSize: 10.5,
                        color: t.fg4,
                        fontFamilyFallback: kMonoFallback)),
            ],
          ),
        ],
      ),
    );
  }
}

File: apps/ios/NanoLink/Screens/AuditLogScreen.swift
import SwiftUI

/// Full audit feed for the active server.
struct AuditLogScreen: View {
    @EnvironmentObject private var store: AppStore
    @Environment(\.nano) private var t
    @State private var loading = true

    var body: some View {
        Group {
            if loading && store.recentActivity().isEmpty {
                ProgressView().controlSize(.large).frame(maxWidth: .infinity, maxHeight: .infinity)
            } else if store.recentActivity().isEmpty {
                empty
            } else {
                ScrollView {
                    LazyVStack(alignment: .leading, spacing: 0) {
                        NanoSectionLabel(tr("audit.entriesCount", ["n": store.recentActivity().count]))
                        NanoCard {
                            ForEach(Array(store.recentActivity().enumerated()), id: \.element.id) { index, entry in
                                row(entry, divider: index < store.recentActivity().count - 1)
                            }
                        }
                    }
                    .padding(EdgeInsets(top: 4, leading: 16, bottom: 50, trailing: 16))
                }
                .refreshable { await load() }
            }
        }
        .background(t.bg.ignoresSafeArea())
        .navigationTitle(tr("audit.title"))
        .navigationBarTitleDisplayMode(.inline)
        .task { await load() }
    }

    private var empty: some View {
        VStack(spacing: 8) {
            Image(systemName: "clock.arrow.circlepath").font(.system(size: 34)).foregroundColor(t.fg4)
            Text(tr("audit.empty")).font(.system(size: 15, weight: .medium)).foregroundColor(t.fg2)
            Text(tr("audit.emptySub")).font(.system(size: 12.5)).foregroundColor(t.fg4)
        }.frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    private func row(_ entry: AuditEntry, divider: Bool) -> some View {
        let tone = entry.ok ? t.ok : t.crit
        let detail = entry.paramsMap.isEmpty
            ? (!entry.target.isEmpty ? entry.target : entry.agentHostname)
            : entry.paramsMap.map { "\($0.key)=\($0.value)" }.sorted().joined(separator: " ")
        return NanoListRow(divider: divider, verticalAlignment: .top) {
            VStack(alignment: .leading, spacing: 3) {
                HStack(spacing: 6) {
                    NanoMono(entry.type, size: 13, color: t.fg, weight: .medium)
                    Text(Fmt.ago(entry.at)).font(.system(size: 11)).foregroundColor(t.fg4)
                }
                if !detail.isEmpty { NanoMono(detail, size: 11.5, color: t.fg3).lineLimit(1) }
                if !entry.error.isEmpty { Text(entry.error).font(.system(size: 11.5)).foregroundColor(t.crit).lineLimit(2) }
                HStack(spacing: 6) {
                    if !entry.user.isEmpty { NanoBadge(text: entry.user, icon: "person") }
                    if !entry.agentHostname.isEmpty { NanoBadge(text: entry.agentHostname, mono: true) }
                    Spacer()
                    if entry.durationMs > 0 { NanoMono("\(entry.durationMs)ms", size: 10.5, color: t.fg4) }
                }.padding(.top, 1)
            }
        } leading: {
            ZStack { Circle().fill(tone.opacity(0.16)); Image(systemName: auditLogIcon(entry.type)).font(.system(size: 15)).foregroundColor(tone) }
                .frame(width: 36, height: 36)
        }
    }

    private func load() async {
        guard let id = store.activeServerId else { loading = false; return }
        await store.fetchRecentActivity(serverId: id, limit: 200)
        loading = false
    }
}

private func auditLogIcon(_ type: String) -> String {
    switch type.lowercased() {
    case "shell.exec", "shell_execute": return "terminal"
    case "service.restart", "docker.restart": return "arrow.clockwise"
    case "system.reboot": return "power"
    default: return "clock.arrow.circlepath"
    }
}
